src/address.rs

```diff
@@ -0,0 +1,247 @@
+extern crate url;
+use self::url::Host;
+extern crate strum;
+use error::SdpParserInternalError;
+use std::convert::TryFrom;
+use std::fmt;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::str::FromStr;
+
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serialize", derive(Serialize))]
+pub enum Address {
+    Fqdn(String),
+    Ip(IpAddr),
+}
+
+impl fmt::Display for Address {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Address::Fqdn(fqdn) => fqdn.fmt(f),
+            Address::Ip(ip) => ip.fmt(f),
+        }
+    }
+}
+
+impl FromStr for Address {
+    type Err = SdpParserInternalError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut e: Option<SdpParserInternalError> = None;
+        if s.find(':').is_some() {
+            match IpAddr::from_str(s) {
+                Ok(ip) => return Ok(Address::Ip(ip)),
+                Err(err) => e = Some(err.into()),
+            }
+        }
+        Host::parse(s)
+            .and_then(|host| match host {
+                Host::Domain(s) => Ok(Address::Fqdn(s)),
+                Host::Ipv4(ip) => Ok(Address::Ip(IpAddr::V4(ip))),
+                Host::Ipv6(ip) => Ok(Address::Ip(IpAddr::V6(ip))),
+            })
+            .map_err(|err| e.unwrap_or_else(|| err.into()))
+    }
+}
+
+impl From<ExplicitlyTypedAddress> for Address {
+    fn from(item: ExplicitlyTypedAddress) -> Self {
+        match item {
+            ExplicitlyTypedAddress::Fqdn { domain, .. } => Address::Fqdn(domain),
+            ExplicitlyTypedAddress::Ip(ip) => Address::Ip(ip),
+        }
+    }
+}
+
+impl PartialEq for Address {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Address::Fqdn(a), Address::Fqdn(b)) => a.to_lowercase() == b.to_lowercase(),
+            (Address::Ip(a), Address::Ip(b)) => a == b,
+            (_, _) => false,
+        }
+    }
+}
+
+#[derive(Clone, Copy, strum_macros::Display, PartialEq, Debug)]
+#[cfg_attr(feature = "serialize", derive(Serialize))]
+pub enum AddressType {
```

Comment:
In the SDP grammar it is called 'addrtype', it does not make the assumption that the value will be an IP version, and is explicitly extensible with via an IANA registry. It does happen that those are the only two we use, but I would rather continue to use the SDP nomenclature.  You know, just in case we ever support Ipx, or DECNet ;).

Reply:
Fixed